### DISCUSSION TITLE
Using mpi4py.futures within Jupyter notebooks

### DISCUSSION THREAD

**Discussion by YarShev** [mpi4py/mpi4py#294] [Q&A] [answered]
Hi there,

I would really like to know how to fix or work around the problem I encountered when running a simple code snipper from the docs in a jupyter notebook.

```python
from mpi4py import MPI
from mpi4py.futures import MPICommExecutor

with MPICommExecutor(MPI.COMM_WORLD, root=0) as executor:
    if executor is not None:
       future = executor.submit(abs, -42)
       assert future.result() == 42
       answer = set(executor.map(abs, [-42, 42]))
       assert answer == {42}
```

When I run this code with

```bash
MPI4PY_FUTURES_MAX_WORKERS=8 mpiexec -n 1 jupyter notebook
```

in a jupyter notebook, I get the error `The kernel appears to have died. It will restart automatically.`. The problem happens on Linux, but on Windows it works fine.

Thanks in advance.

**Comment by dalcinl** [comment]
I see you are setting `MPI4PY_FUTURES_MAX_WORKERS=8`. But then  executing `jupyter` with just one MPI process. Therefore, you should not use `MPICommExecutor`, you should use instead `MPIPoolExecutor` to get `mpi4py.futures` spawn the 8 worker processes you want.

Using mpi4py under Jupyter is a bit tricky, mostly because MPI implementations sometimes do not support spawning new processes (the 8 workers you want) if the parent process (the kernel) was not executed via `mpiexec`. Note that you are using `mpiexec` indeed, but you are using it to start Jupyter, and that does not mean Jupyter will start the kernel process with mpiexec. Things may not work.

Alternatively, you may setup Jupyter to start a cluster using the MPI engine. This cluster needs just one process (the parent), then `mpi4py.futures` will take care of spawning the workers via standard MPI mechanisms. Check Jupyter docs about all that stuff. Unfortunately, I'm not a savvy user of parallel IPython/Jupyter, sorry I cannot provide more useful help.

**Comment by YarShev** [comment]
Oh, now I see the difference between `MPICommExecutor` and `MPIPoolExecutor`. The former just splits the set of MPI processes within a MPI communicator in one master process and many workers processes. However, I am still experiencing the same behavior even with `MPIPoolExecutor` on both Linux (it doesn't work) and Windows (it works). I tried this script.

```python
from mpi4py import MPI
from mpi4py.futures import MPIPoolExecutor

if __name__ == "__main__":
    with MPIPoolExecutor() as executor:
        if executor is not None:
           future = executor.submit(abs, -42)
           assert future.result() == 42
           answer = set(executor.map(abs, [-42, 42]))
           assert answer == {42}
```

Why doesn't the use of mpiexec mean Jupyter will start the kernel process with mpiexec?

Regarding the set up of jupyter to start a cluster using the MPI engine, I will try to find the docs and go through it.

Meanwhile, could you please tag some more people who may be aware of this problem and of its resolving?

**Comment by dalcinl** [comment]
> Why doesn't the use of mpiexec mean Jupyter will start the kernel process with mpiexec?

If I understand things correctly, the kernel runs in a separate process created at some point, and I don't think that process is under mpiexec control. 

> Meanwhile, could you please tag some more people who may be aware of this problem and of its resolving?

I don't really know. Maybe @leofang can think of someone?

PS: What MPI implementation are you using on Linux and Windows?

**Comment by leofang** [comment]
>> Regarding the set up of jupyter to start a cluster using the MPI engine, I will try to find the docs and go through it.

IIUC you can refer to these resources (disclaimer: I am on my cell and I didn't try)
- https://ipyparallel.readthedocs.io/en/latest/reference/mpi.html
- https://charlesreid1.com/wiki/Jupyter/MPI#Start_MPI_Cluster

I think the key is to ensure whenever Jupyter launches a kernel, it's launched under mpiexec. Then the kernel can spawn new processes as long as your MPI backend supports it.

> > Meanwhile, could you please tag some more people who may be aware of this problem and of its resolving?
> 
> I don't really know. Maybe @leofang can think of someone?

 IIRC @minrk is the expert? 🙂

**Comment by YarShev** [comment]
> > Why doesn't the use of mpiexec mean Jupyter will start the kernel process with mpiexec?
> 
> If I understand things correctly, the kernel runs in a separate process created at some point, and I don't think that process is under mpiexec control.

That might be the case because I see three new processes (mpiexec, hydra_pmi_proxy, jupyter-noteboo) in top on Linux and two new processes (mpiexec.exe, jupyter.exe, but there is no hydra_pmi_proxy.exe) in tasklist on Windows when I run.

```bash
mpiexec -n 1 jupyter notebook
```

> PS: What MPI implementation are you using on Linux and Windows?

**Windows**
```bash
mpi                       1.0                       msmpi    conda-forge
mpi4py                    3.1.4            py38h73c715a_0    conda-forge
msmpi                     10.1.1               h3502643_7    conda-forge
```

**Linux**
```bash
mpi                       1.0                       mpich    conda-forge
mpi4py                    3.1.4            py38h97ac3a3_0    conda-forge
mpich                     4.0.3              h846660c_100    conda-forge
```

**Comment by YarShev** [comment]
> > > Regarding the set up of jupyter to start a cluster using the MPI engine, I will try to find the docs and go through it.
> 
> IIUC you can refer to these resources (disclaimer: I am on my cell and I didn't try)
> 
> * https://ipyparallel.readthedocs.io/en/latest/reference/mpi.html
> * https://charlesreid1.com/wiki/Jupyter/MPI#Start_MPI_Cluster
> 
> I think the key is to ensure whenever Jupyter launches a kernel, it's launched under mpiexec. Then the kernel can spawn new processes as long as your MPI backend supports it.

I will go through these resources. They probably explain how to configure jupyter in order for it to run a kernel under mpiexec.

> IIRC @minrk is the expert? 🙂

I would be really appreciate any help 🙂.

**Comment by dalcinl** [comment]
Oh, I see... you are using MPICH. That explains why your kernel process dies. As the kernel process is not executed with `mpiexec`, the MPI spawn call to create worker process when using `MPIPoolExecutor` fails. That's a know issue with the MPICH version you are using. Fortunately, next MPICH release 4.1 should work just fine.

@hzhou This is precisely the use case I've been so insistent in the past about getting singleton init + spawn working seamlessly. Thanks again for your prompt action on fixing this long standing issue.

**Reply by YarShev** [reply]
@dalcinl, do you mean that starting MPICH release 4.1 it should be able to run mpi4py in a jupyter notebook with the command I posted and without any additional setup?

**Reply by dalcinl** [reply]
Yes, exactly. Actually, you will not even need to execute `jupyter` with `mpiexec`. Things  should just work out of the box.

**Reply by YarShev** [reply]
That would be great! Look forward to trying the release 4.1.

**Comment by minrk** [comment, marked as answer]
> I think the key is to ensure whenever Jupyter launches a kernel, it's launched under mpiexec

To start a kernel process with mpiexec, you'll want to modify `kernel.json` and add `mpiexec` to the command launched (`argv`). You can locate the kernelspec directory with `jupyter kernelspec list`, but it will generally be at `$PREFIX/share/jupyter/kernels/python3/kernel.json`. You can change the command to be a more customizable launch script, rather than launching ipykernel directly, if you want to be able to e.g. try to _detect_ whether mpiexec should be invoked or not.

You may also want to _copy_ the `python3` kernel to a new directory, e.g. `mpi` (this will create an 'mpi' kernel in your kernel list), if you want to be able to pick mpi or not on a per-notebook basis.

**Reply by YarShev** [reply]
@minrk, I modified `kernel.json` by adding "mpiexec", "-n", "1" to `args`. Now the example just hangs. Am I missing out on something?

**Reply by YarShev** [reply]
> You may also want to copy the python3 kernel to a new directory, e.g. mpi (this will create an 'mpi' kernel in your kernel list), if you want to be able to pick mpi or not on a per-notebook basis.

Also, I am not sure that I fully get the above. Could you please elaborate a bit? How is an 'mpi' kernel able to be created automatically just by copying python3 kernel to `mpi` directory?

**Reply by dalcinl** [reply]
@YarShev For example, I installed jupyter with `python -m pip install jupyter`. Then the prefix directory is `~/.local`.
Therefore, I have this file:

```
$ cat ~/.local/share/jupyter/kernels/python3/kernel.json 
{
 "argv": [
  "python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3 (ipykernel)",
 "language": "python",
 "metadata": {
  "debugger": true
 }
```
If you add `"mpiexec", "-n", "1"` at the beginning of the list, then you get what you need.

But even better, you can make a copy of the `python3` folder:
```
cp -r ~/.local/share/jupyter/kernels/python3 ~/.local/share/jupyter/kernels/python3mpi
```
and then edit `~/.local/share/jupyter/kernels/python3mpi/kernels.json` to add the `mpiexec` command as explained above, but also change `display_name` to let say `"Python 3 (ipykernel) with MPI"`. That way you can specifically select the Python kernel with MPI-enabled using the dropdown menu in your browser.

**Reply by YarShev** [reply]
@dalcinl, @minrk, sorry guys for the confusion. I tried Intel MPI implementation and the example hangs with it. Everything is fine with MPICH.

**Reply by YarShev** [reply]
@dalcinl, thanks for the explanation, now it is clear to me. Btw, do you experience the issue with Intel MPI (`conda install mpi4py -c intel`)?

**Reply by dalcinl** [reply]
Just [a few issues](https://community.intel.com/t5/Intel-oneAPI-HPC-Toolkit/Issues-running-mpi4py-testsuite-on-GitHub-Actions/m-p/1428130), although I believe mine are unrelated to yours. I may be wrong, though.

**Reply by YarShev** [reply]
I mean the example here if you run it in jupyter.

**Reply by YarShev** [reply]
@dalcinl. @minrk, I am also wondering why the things work on Windows without editing `kernel.json`?

**Reply by dalcinl** [reply]
@YarShev To test on my side I would need a full instructions about how to reproduce the conda environment, including the base installer you used for yor conda-based installation (Anaconda? Miniconda? Miniforge?). Things may work on Windows because MPI spawn with Intel MPI simply works out of the box in your system, without using `mpiexec`. You first told us that you were using MPICH, but now you are telling me that the issues are with Intel MPI. If you change the backend MPI, things change. If you try to use Open MPI, you will have trouble again. All these issues are not mpi4py's fault, so there is not very much we can do to help you beyond telling you to try using `mpiexec` to launch the kernel.

**Reply by YarShev** [reply]
Yes, I have not tested the things with Open MPI on Windows. It may not work. I got that the issues come from a concrete MPI implementation. Thanks!

**Reply by YarShev** [reply]
For you to reproduce:

- I installed `Miniconda3-py39_4.12.0-Linux-x86_64.sh`
- created env with `conda create -n test-intel-mpi python=3.8`
- activated env `conda activate test-intel-mpi`
- installed Intel MPI `conda install mpi4py -c intel`
- installed jupyter `pip install jupyter`

Hope it is sufficient.

**Reply by dalcinl** [reply]
Your example deadlocks. It happens at `executor.shutdown()`, which is triggered by the `with` statement.  The problem is most likely related to issue #15. [This hacks](https://github.com/mpi4py/mpi4py/issues/15#issuecomment-765658147) works, but it requires modifying the installed mpi4py code.

**Reply by dalcinl** [reply]
After upgrading to Py 3.10 and latest mpi4py 3.1.4, things get even worse, the `recv_mprove` hack no longer works. Long story short, IMHO, Intel MPI is broken. As an occasional user with no access to Intel MP source code, I'm not in a position to investigate this further.

**Reply by YarShev** [reply]
Thank you very much for the investigation! This is very helpful for me to move forward. Could you please elaborate the issue with `rc.recv_mprobe` or point me out at some reference discussion? Say If I want to use MPI probing in terms of performance, why I can't use it with mpi4py (I have to disable it to make things work)?

**Reply by YarShev** [reply]
Btw, I see only comm.Probe method in the mpi4py documentation - https://mpi4py.readthedocs.io/en/stable/reference/mpi4py.MPI.Comm.html?highlight=probe#mpi4py.MPI.Comm.Probe. Why there is no like comm.MProbe as in C?

**Reply by dalcinl** [reply]
> Could you please elaborate the issue with rc.recv_mprobe or point me out at some reference discussion?

Intel had a [webpage](https://software.intel.com/en-us/articles/python-mpi4py-on-intel-true-scale-and-omni-path-clusters), but looks like it is gone by now. That page simply mentioned the workaround I explained above, but the Intel folks never really addressed the root of the problem. Many years after, I'm are still there advising people about it, it is really frustrating.

> why I can't use it with mpi4py (I have to disable it to make things work)?

It is not about performance. The actual problem is that some codes eventually deadlock.

> Why there is no like comm.MProbe as in C?

Simply because there is no 'MProbe' in C, but `Mprobe` instead. Just look in the webpage you linked for the entry right above `Probe`.

PS: I converted your original issue as a conversation to mark this answer as accepted. Your original question was about MPICH. Now we are digressing to address Intel MPI. We should stop adding to this thread to avoid confusion for future readers. Feel free to open a new issue, or let's continue via email in mpi4py@googlegroups.com

**Reply by YarShev** [reply]
Yes, I think we duscussed enough here. Thank you very much.
